design/notifications.md

```diff
@@ -0,0 +1,405 @@
+the next one to arrive. If it decides that one or more users have to be informed of the event then processing will
+continue. This decision will be influenced by the category and the update type, so each handler will likely make this
+decision in a unique manner except. In the case of the backwards compatible mode, the decision will be made in the same
+way as in the curernt notification agent out of necessity.
+
+The next step is to determine who should receive the notifications. This is another decision that is likely to vary
+depending on category and update type. The handler for backwards compatible notifications will only send the
+notification to the user mentioned in the `user` field of the notification request.
+
+Next, the event recorder will have to determine whether or not an email should be sent. And once again, this decision
+will probably vary depending on the category and update type. Eventually, we'll probably also want to have a
+subscription mechanism so that users can choose which types of notifications they get emails for. For the backwards
+compatibility case, however, the user will only receive an email if the notificaton request indicates that an email
+should be sent.
+
+The next step is to determine exactly what each message should say. At the risk of sounding repetitive, this is
+something that will vary depending on the category and update type. We may want to explore internationalization and
+localization for our notification system in the near future, too. The current system, where each service determines the
+wording to use for each type of notification, doesn't really lend itself to internationalization and localization, so
+this won't be implemented right away. We may want to investigate adding `go-i18n` (assuming we use Go) to one or more
+notification services when we start adding handlers other than the backwards compatible handler, though.
+
+Another thing to consider is how we're going to ensure message order. We can't really provide guarantees for the order
+in which messages are processed, but we can take some steps to control the order in which messages are displayed to the
+user. To this end, each update message will have a timestamp associated with it when it's posted to the AMQP
+exchange. This timestmap will be recorded in the notification database, and it will be used to determine notification
+order when requests are made to list notifications. Current notification requests don't have a timestamp associated with
+them because they're recorded using synchronuous HTTP calls. Because of this, the endpoint in the service that provides
+the backwards compatible REST API for notifications will add the timestamp to the notification request before publishing
+it to the exchange. For updates that are posted directly to AMQP, the service itself will be responsible for including
+the timestamp in the request.
```

Comment:
There's some amount of established (in the AMQP world, not ours) practice of adding timestamps as message headers to AMQP messages. However, I'm mentioning it here in part because there's an issue -- as I recall, the precision on those timestamps is somewhat low (I forget if it's seconds, perhaps, or milliseconds, where a lot of our own stuff we use nanoseconds). So this probably doesn't change anything immediately, but I recommend not using that field (or using it in addition to a more precise timestamp in the message itself).